add ReadLogin method to parse TDS7 login packet

doodlesbykumbi opened the pull request.

jonahx commented on tds.go:
```diff
@@ -465,6 +465,133 @@ func sendLogin(w *TdsBuffer, login login) error {
+// changes to code.
+type Login struct {
+	login
+}
```
I like this hack!

diverdane replied:
Will this actually expose `login` (and its fields) to external packages? The embedded `login` will remain unexported, and as far as I can tell, it won't be exposed. Or is the intent to return `Login` as an opaque object, without access to private fields?

doodlesbykumbi replied:
@diverdane The public fields on an otherwise private type (login) are public because login is embedded in Login, login's fields are elevated to being fields on Login. It just works

diverdane replied:
@doodlesbykumbi Slickness!!!!! Thanks for the explanation.

jonahx commented on tds.go:
```diff
@@ -465,6 +465,133 @@ func sendLogin(w *TdsBuffer, login login) error {
+
+	// NOTE: offsets are calculated from after the packet header (not the login header)
+	// it is for this reason that the we must add headerSize to all the offsets to get
+	// the correct position within the buffer
```
We could avoid the duplication with  a local helper function:
```go
	position := func(offset uint16) int {
		return headerSize + int(offset)
	}
```
Used like:
```go
r.rpos = position(hdr.HostNameOffset)
```

jonahx commented on tds_test.go:
```diff
@@ -70,6 +70,206 @@ func TestSendLogin(t *testing.T) {
+		fmt.Println("TDSVersion")
+		fmt.Println("Expected:", originalLogin.TDSVersion)
+		fmt.Println("Returned:", login.login.TDSVersion)
+		t.Error("input output don't match for TDSVersion")
```
Assuming we can't figure out how to make deepEqual work (I think we can) we could at least package up these 4 repeated test assertions into a local func.

doodlesbykumbi replied:
I’m thinking to stick with the lengthy 1 by 1 assertions because when there’s a failure we have insight into what’s wrong, less so with DeepEqual.

diverdane replied:
Curious... Did you have a chance to figure out why DeepEqual wasn't working in this case, or is this a `don't care` now?

doodlesbykumbi replied:
@diverdane It is because one of the fields is a byte slice. Explanation straight from the Go docs https://golang.org/pkg/reflect/#DeepEqual
> Note that a non-nil empty slice and a nil slice (for example, []byte{} and []byte(nil)) are not deeply equal.

jonahx replied:
@doodlesbykumbi That repetition is too much.  How about you just use:

https://godoc.org/github.com/stretchr/testify/assert#EqualValues

This uses `DeepEqual` underneath, but also performs and diff and reports that when the match fails.

doodlesbykumbi replied:
That'd require introducing a new dep to a fork. For only addressing this particular test case I don't think we'd be justified. This repo has minimal external deps and I think for now it's reasonable to stick with that

doodlesbykumbi replied:
@jonahx That'd require introducing a new dep to a fork. For only addressing this particular test case I don't think we'd be justified. This repo has minimal external deps and I think for now it's reasonable to stick with that

jonahx replied:
Yeah, I can see your point there.  How about just a helper func then to reduce the boilerplate:

```go
func Fail(t testing.T, name string, expected, actual interface{}) {
	fmt.Println("Expected:", expected)
	fmt.Println("Returned:", actual)
	t.Errorf("input output don't match for %s", name)
}
```

doodlesbykumbi committed:
add ReadLogin method to parse TDS7 login packet

diverdane commented on tds.go:
```diff
@@ -465,6 +465,134 @@ func sendLogin(w *TdsBuffer, login login) error {
+		return nil, err
+	}
+
+	r.rpos = offsetAfterHeader(hdr.HostNameOffset)
```
*Probably* not worth it, but you could possibly eliminate most of the `r.rpos = offsetAfterHeader(***)` lines by creating a helper function:
```
func readUcs2WithOffset(r io.Reader, numchars int, offset uint16) (res string, err error) {
    r.pos = offsetAfterHeader(offset)
    return readUcs2(r, numchars)
}
```
where the calls would be e.g.:
```
    hostname, err := readUcs2WithOffset(r, int(hdr.HostnameLength), hdr.HostNameOffset)
```
But it's 6 o' one, half-dozen of another.

jonahx replied:
@diverdane Correction: _Definitely_ worth it :)

diverdane reviewed (commented):
All looks great (agree with Jonah's suggestions), thanks for digging into this login protocol!

BradleyBoutcher committed:
Add an 'if' check to prevent null pointer exceptions

If we test the mssql db plugin without secretless, no channel will
be passed in through the context. To avoid any issues testing the
plugin, we want to check here.

BradleyBoutcher committed:
Merge branch 'master' into 975-mssql-hardcoded-responses
